File: pipeline/submerged.py
"""Publish the submerged-landscape assessment from its source registers."""
import hashlib
import json
from pathlib import Path
import struct
import re
from urllib.parse import urlsplit
try:
    from pipeline.catalogue import esc, a, shell
except ModuleNotFoundError:
    from catalogue import esc, a, shell

ROOT=Path(__file__).resolve().parents[1]
DOCS=ROOT/'docs'


def validate(data,manifest):
    source_ids={s['id'] for s in data['sources']}
    if len(source_ids)!=len(data['sources']):raise ValueError('Duplicate submerged source')
    grids={g['id']:g for g in manifest['grids']}
    if len(grids)!=len(manifest['grids']) or 'world' not in grids:raise ValueError('Invalid grid identifiers')
    for s in data['sources']:
        if urlsplit(s['url']).scheme!='https' or not urlsplit(s['url']).netloc:raise ValueError('Unsafe source URL')
        if not s.get('checked') or not s.get('note'):raise ValueError('Incomplete submerged source')
    for key in ['projects','datasets']:
        if len({v['id'] for v in data[key]})!=len(data[key]):raise ValueError('Duplicate submerged record')
        for r in data[key]:
            if not re.fullmatch(r'[a-z0-9]+(?:-[a-z0-9]+)*',r['id']):raise ValueError('Unsafe record identifier')
            if not r['sources'] or not set(r['sources'])<=source_ids:raise ValueError('Unsupported submerged record')
            if not r.get('limit'):raise ValueError('Missing evidence qualification')
    for p in data['projects']:
        if p['region'] not in grids or not (-90<=p['lat']<=90 and -180<=p['lon']<=180):raise ValueError('Invalid project location')
    for g in manifest['grids']:
        source=(ROOT/g['source_path']).resolve();output=(DOCS/g['path']).resolve()
        if not source.is_relative_to(ROOT/'data/submerged') or not output.is_relative_to(DOCS/'data/submerged'):raise ValueError('Unsafe relief path')
        raw=source.read_bytes();binary=output.read_bytes()
        if hashlib.sha256(raw).hexdigest()!=g['source_sha256'] or hashlib.sha256(binary).hexdigest()!=g['sha256']:
            raise ValueError('Relief data changed')
        if len(binary)!=g['width']*g['height']*2:raise ValueError('Incorrect relief grid dimensions')
        if not (-180<=g['west']<g['east']<=180 and -90<=g['south']<g['north']<=90):raise ValueError('Invalid relief extent')


def preview(g,depth,projects=()):
    values=struct.unpack('<'+'h'*(g['width']*g['height']),(DOCS/g['path']).read_bytes())
    paths={1:[],2:[]}
    def classify(z):return 0 if z==g['nodata'] else 1 if z>=0 else 2 if z>=-depth else 0
    for y in range(g['height']):
        row=values[y*g['width']:(y+1)*g['width']]
        start=0;cat=classify(row[0])
        for x in range(1,g['width']+1):
            nxt=classify(row[x]) if x<g['width'] else -1
            if nxt!=cat:
                if cat:paths[cat].append(f'M{start} {y}h{x-start}v1H{start}z')
                start=x;cat=nxt
    markers=[]
    for i,p in enumerate(projects):
        if not (g['west']<=p['lon']<=g['east'] and g['south']<=p['lat']<=g['north']):continue
        x=(p['lon']-g['west'])/(g['east']-g['west'])*(g['width']-1)
        y=(g['north']-p['lat'])/(g['north']-g['south'])*(g['height']-1)
        markers.append(f'<g><title>{esc(p["name"])} · {esc(p["precision"])}</title><circle cx="{x:.1f}" cy="{y:.1f}" r="12.7" fill="#f3ecda" stroke="#173b45"/><text x="{x:.1f}" y="{y+5:.1f}" text-anchor="middle" font-family="monospace" font-size="16" fill="#173b45">{i+1}</text></g>')
    return f'''<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 {g['width']} {g['height']}" role="img"><title>NOAA ETOPO1 modern relief with elevations between minus {depth} metres and zero highlighted</title><desc>Modern land in light green, shallow negative elevations in gold, deeper water in blue. This is a depth screen, not a reconstruction of a dated ancient coastline. Inland depressions are not excluded. Numbered markers are approximate project locations.</desc><rect width="100%" height="100%" fill="#173b45"/><path fill="#d6d4b9" d="{''.join(paths[1])}"/><path fill="#d4a763" d="{''.join(paths[2])}"/>{''.join(markers)}</svg>'''


def page(data,manifest):
    sources={s['id']:s for s in data['sources']}
    def refs(ids):return ' · '.join(a(sources[id]['url'],sources[id]['title']) for id in ids)
    projects=''.join(f'''<article class="water-project" id="project-{p['id']}"><div class="project-number">{i+1:02d}</div><div><div class="micro">{esc(p['status'])}</div><h3>{esc(p['name'])}</h3><p class="project-lead">{esc(p['label'])}</p><p>{esc(p['finding'])}</p><details><summary>Evidence, limits &amp; next investigation</summary><dl><dt>Programme / record</dt><dd>{esc(p['activity'])}</dd><dt>What remains open</dt><dd>{esc(p['limit'])}</dd><dt>A useful next step</dt><dd>{esc(p['next'])}</dd><dt>Map precision</dt><dd>{esc(p['precision'])}</dd></dl></details><p class="water-sources">{refs(p['sources'])}</p><a class="project-map-link" href="submerged.html?region={p['region']}&amp;depth=60#landscape" data-region-link="{p['region']}">Explore the regional depth view ↗</a></div></article>''' for i,p in enumerate(data['projects']))
    datasets=''.join(f'''<article class="water-dataset" id="dataset-{d['id']}"><div class="micro">{esc(d['coverage'])}</div><h3>{esc(d['name'])}</h3><p>{esc(d['use'])}</p><details><summary>Resolution, access &amp; limitations</summary><dl><dt>Resolution</dt><dd>{esc(d['resolution'])}</dd><dt>Access checked</dt><dd>{esc(d['access'])}</dd><dt>Limits</dt><dd>{esc(d['limit'])}</dd></dl></details><p class="water-sources">{refs(d['sources'])}</p></article>''' for d in data['datasets'])
    source_html=''.join(f'<li><span class="micro">{esc(s["kind"])}</span><h3>{a(s["url"],s["title"])}</h3></li>' for s in data['sources'])
    locations=''.join(f'<a href="#project-{p["id"]}" data-project-lat="{p["lat"]}" data-project-lon="{p["lon"]}" data-project-number="{i+1}"><span>{i+1:02d}</span> {esc(p["name"])}</a>' for i,p in enumerate(data['projects']))
    body=f'''<main id="main"><header class="water-hero"><div class="cat-wrap"><div class="micro">A second inquiry · Submerged landscapes</div><h1>Another world<br><em>lies offshore.</em></h1><div class="water-intro"><p>The sea covered coastlines, rivers and places people once lived.</p><p>What survives—and what can today’s data reveal?</p></div><nav aria-label="In this programme"><a href="#landscape">Explore the shelves ↓</a><a href="#experiment">The first data test</a><a href="#projects">Discoveries &amp; projects</a><a href="#data">Open data</a><a href="#programme">A research programme</a></nav></div></header>
<section id="landscape" class="water-explorer cat-wrap"><div class="section-top"><div><div class="micro">Begin with the shape of the land</div><h2>Look beyond today’s coast.</h2></div><p>Modern relief · depth experiment.<br><b>No ancient date assigned.</b></p></div>
<div class="water-tools" id="water-tools" hidden><div class="region-buttons" aria-label="Map region">{''.join(f'<button type="button" data-water-region="{g["id"]}" aria-pressed="{str(g["id"]=="world").lower()}">{esc(g["label"])}</button>' for g in manifest['grids'])}</div><label class="depth-control" for="water-depth">Highlight down to <output id="depth-output">60 m</output><input type="range" id="water-depth" min="0" max="120" step="5" value="60" aria-describedby="depth-explanation"></label></div>
<div class="water-map-pair"><figure><div class="map-overline">Modern elevation baseline</div><canvas id="modern-canvas" hidden role="img" aria-label="Modern relief map"></canvas><img class="water-fallback" id="modern-fallback" src="img/submerged/world-modern.svg" alt="World relief from NOAA ETOPO1; present non-negative elevations in light green, negative elevations in blue"><figcaption>Present relief · an older global compilation</figcaption></figure><figure><div class="map-overline">Depth screen · <span id="scenario-title">0 to −60 m</span></div><canvas id="depth-canvas" hidden role="img" aria-label="Relief map with the selected shallow depth interval highlighted"></canvas><img class="water-fallback" id="depth-fallback" src="img/submerged/world-depth.svg" alt="The same world relief with elevations from minus 60 metres to zero highlighted in gold; no date assigned"><figcaption>Gold = cells within the selected depth interval</figcaption></figure></div>
<div class="water-map-legend"><span>Modern non-negative elevations</span><span>Selected shallow elevations</span><span>Deeper water</span></div><p id="water-status" role="status" class="water-map-status">World view · static 60 m depth screen. Interactive controls require JavaScript.</p><nav class="water-project-locations" aria-label="Projects in the mapped region">{locations}</nav><p class="water-map-status">Markers locate project regions. Map areas are not directly comparable.</p><p id="depth-explanation" class="depth-explanation">Gold marks a depth interval on <b>modern relief</b>, including inland depressions. Land motion, sediment, erosion and past ice cover are not modelled. It is not an ancient coastline.</p><details class="water-method"><summary>What would a map of 12,000 years ago require?</summary><p>A dated regional sea-level model and a reconstructed former land surface, with uncertainty. The Last Glacial Maximum was earlier than 12,000 years ago; those shorelines differ.</p><p>North Sea peat records offer a starting point. “12,000 years BP” means before 1950, approximately 10,050 BCE; inundation dates vary by place.</p><p class="water-sources">{refs(['lambeck','northsea-peats','northsea-data'])}</p><p class="water-sources">Prototype: NOAA ETOPO1 (2009), subsampled at 20′ globally and 2′–6′ regionally. <a href="data/submerged/grids.json">Exact requests, extents, hashes and limitations</a>. Source NetCDF files are retained in the repository.</p></details><noscript><p class="water-map-status">The maps above remain readable without scripting. All projects, datasets and sources follow below.</p></noscript></section>
<section id="scope" class="water-scope cat-wrap"><div><div class="micro">The initial assessment</div><h2>The missing landscape is real.<br>Its contents are a research question.</h2></div><div><p>Documented submerged sites justify the search. How much evidence survives, and what it contains, remain open questions. {refs(['pavlopetri','murujuga'])}</p><p>Start where <b>past habitability, preservation and usable surveys</b> overlap.</p><p class="water-sources">Assessment updated {data['edition']}.</p></div></section>
<section id="experiment" class="water-scope cat-wrap"><div><div class="micro">New · a test on actual survey data</div><h2>What a coarse grid erases.</h2></div><div><p>The Baltic alignment’s 0.5 m survey lets us measure how averaging removes fine relief. At 115 m, this test retains just <b>0.3% of local relief variance</b>.</p><p>A data-resolution result, not an AI detection or a wall-recovery score.</p><a class="quiet-action" href="bathymetry-lab.html">Inspect the survey and measurements ↗</a></div></section>
<section id="projects" class="water-section cat-wrap"><div class="section-top"><div><div class="micro">The field, in six starting places</div><h2>What has come into view?</h2></div><p>Discoveries, disputed claims and landscapes to investigate.</p></div><div class="water-projects">{projects}</div></section>
<section id="data" class="water-section water-data-section"><div class="cat-wrap"><div class="section-top"><div><div class="micro">An inventory for investigation</div><h2>Data we can begin with.</h2></div><p>Sources, access and resolution, checked before analysis.</p></div><div class="water-data-grid">{datasets}</div></div></section>
<section id="programme" class="water-section cat-wrap"><div class="section-top"><div><div class="micro">The research programme</div><h2>Two practical starting points.</h2></div><p>North Sea landscape reconstruction.<br>Baltic feature detectability.</p></div><ol class="water-steps"><li><b>Reconstruct a dated landscape.</b><p>Combine regional sea-level records, land motion and former surfaces. Keep uncertainty visible.</p></li><li><b>Map where the evidence survives.</b><p>Check survey coverage, sediment burial and resolution before interpreting an empty map.</p></li><li><b>Benchmark the observer.</b><p>Use labelled structures and natural alternatives from separate surveys. Compare AI with simple detection methods.</p></li><li><b>Test candidate sites independently.</b><p>Prioritise freshwater margins and preserved surfaces, then validate with complementary surveys and archaeology.</p></li></ol><div class="water-programme-links"><a href="data/next-tests.md">Concrete next tests ↗</a><a href="https://github.com/ejhong/birdmen/blob/main/research/submerged/PROGRAMME.md">Full programme ↗</a><a href="data/submerged/register.json">Project &amp; data register ↗</a></div></section>
<section id="sources" class="sources-section cat-wrap"><div class="section-top"><div><div class="micro">Follow the evidence</div><h2>Sources for this assessment.</h2></div><p>Source notes and checks: <a href="data/submerged/register.json">full register</a>.</p></div><ol class="dossier-sources">{source_html}</ol></section></main>'''
    result=shell('Submerged worlds','Explore drowned landscapes, underwater archaeology, open bathymetry and a new research programme.',body,extra='<link rel="stylesheet" href="submerged.css"><script type="module" src="submerged.js"></script>',current='submerged.html')
    return result.replace('class="catalogue-site"','class="catalogue-site submerged-site"').replace('href="catalogue.html#about"','href="submerged.html#scope"').replace('href="data/catalogue.json"','href="data/submerged/register.json"')


def outputs():
    data=json.loads((ROOT/'research/submerged/register.json').read_text())
    manifest=json.loads((ROOT/'research/submerged/grids.json').read_text())
    validate(data,manifest)
    world=next(g for g in manifest['grids'] if g['id']=='world')
    return {DOCS/'submerged.html':page(data,manifest),DOCS/'data/submerged/register.json':json.dumps(data,ensure_ascii=False,indent=2)+'\n',DOCS/'data/submerged/grids.json':json.dumps(manifest,indent=2)+'\n',DOCS/'img/submerged/world-modern.svg':preview(world,0,data['projects']),DOCS/'img/submerged/world-depth.svg':preview(world,60,data['projects'])}

File: docs/submerged.js
import {readWaterState,classifyElevation,decodeGrid} from './submerged-model.mjs';
const $=s=>document.querySelector(s);
let manifest,state,current,loadVersion=0;
const cache=new Map();
const locations=[...document.querySelectorAll('[data-project-lat]')].map(link=>({
  link,lat:+link.dataset.projectLat,lon:+link.dataset.projectLon,number:link.dataset.projectNumber
}));
const inGrid=(p,g)=>p.lon>=g.west&&p.lon<=g.east&&p.lat>=g.south&&p.lat<=g.north;
const coordinate=(n,positive,negative)=>`${Math.abs(n).toFixed(1)}°${n<0?negative:positive}`;
function paint(canvas,grid,values,depth) {
  canvas.width=grid.width;canvas.height=grid.height;
  const ctx=canvas.getContext('2d'),image=ctx.createImageData(grid.width,grid.height);
  for(let i=0;i<values.length;i++) {
    const z=values[i],category=classifyElevation(z,depth,grid.nodata);
    let rgb;
    if(category==='land') {const h=Math.min(1,Math.max(0,z)/3500);rgb=[214-h*47,212-h*43,185-h*35];}
    else if(category==='shelf') rgb=[212,167,99];
    else if(category==='missing') rgb=[120,120,120];
    else {const h=Math.min(1,Math.abs(z)/4500);rgb=[36-h*18,79-h*29,86-h*27];}
    image.data.set([...rgb.map(Math.round),255],i*4);
  }
  ctx.putImageData(image,0,0);
  const radius=Math.max(4,grid.width/85);
  for(const p of locations.filter(p=>inGrid(p,grid))) {
    const x=(p.lon-grid.west)/(grid.east-grid.west)*(grid.width-1),y=(grid.north-p.lat)/(grid.north-grid.south)*(grid.height-1);
    ctx.beginPath();ctx.arc(x,y,radius,0,Math.PI*2);ctx.fillStyle='#f3ecda';ctx.fill();
    ctx.lineWidth=Math.max(.6,grid.width/1400);ctx.strokeStyle='#173b45';ctx.stroke();
    ctx.fillStyle='#173b45';ctx.font=`${radius*1.25}px monospace`;ctx.textAlign='center';ctx.textBaseline='middle';ctx.fillText(p.number,x,y+.1);
  }
  // Coordinate bounds convey location without inventing ancient place names.
  ctx.textAlign='left';ctx.textBaseline='alphabetic';
  ctx.fillStyle='#f4eee1';ctx.font=`${Math.max(7,grid.width/90)}px monospace`;
  ctx.fillText(`${coordinate(grid.west,'E','W')}–${coordinate(grid.east,'E','W')} · ${coordinate(grid.south,'N','S')}–${coordinate(grid.north,'N','S')}`,grid.width*.025,grid.height*.955);
  canvas.hidden=false;
  canvas.setAttribute('aria-label',`${grid.label}. ${depth?`Elevations from zero to minus ${depth} metres highlighted in gold. No date assigned.`:'Modern relief baseline.'}`);
}
function draw() {
  if(!current)return;
  paint($('#modern-canvas'),current.grid,current.values,0);
  paint($('#depth-canvas'),current.grid,current.values,state.depth);
  $('#modern-fallback').hidden=true;$('#depth-fallback').hidden=true;
  $('#depth-output').value=`${state.depth} m`;$('#scenario-title').textContent=`0 to −${state.depth} m`;
  $('#water-depth').value=state.depth;
  for(const p of locations)p.link.hidden=!inGrid(p,current.grid);
  const spacing=Math.round(current.grid.step_degrees*60);
  $('#water-status').textContent=`${current.grid.label} · ${spacing} arc-minute sampling · NOAA ETOPO1 (2009). Depth screen: ${state.depth} m below the model’s mean sea level. No ancient date assigned.`;
  document.body.dataset.waterReady='true';
  for(const b of document.querySelectorAll('[data-water-region]'))b.setAttribute('aria-pressed',b.dataset.waterRegion===state.region);
}
async function loadRegion() {
  const version=++loadVersion,grid=manifest.grids.find(g=>g.id===state.region);
  $('#water-status').textContent=`Loading ${grid.label}…`;
  $('#water-depth').disabled=true;
  for(const b of document.querySelectorAll('[data-water-region]'))b.setAttribute('aria-pressed',b.dataset.waterRegion===state.region);
  try {
    if(!cache.has(grid.id))cache.set(grid.id,fetch(grid.path).then(r=>{if(!r.ok)throw Error('Relief unavailable');return r.arrayBuffer();}).then(buffer=>decodeGrid(buffer,grid)));
    const values=await cache.get(grid.id);if(version!==loadVersion)return;
    current={grid,values};draw();$('#water-depth').disabled=false;
  } catch {
    if(version!==loadVersion)return;
    cache.delete(grid.id);current=null;
    $('#modern-canvas').hidden=true;$('#depth-canvas').hidden=true;
    $('#modern-fallback').hidden=false;$('#depth-fallback').hidden=false;
    for(const p of locations)p.link.hidden=false;
    $('#scenario-title').textContent='0 to −60 m · static world fallback';
    $('#water-status').textContent='This regional grid could not load. Showing the static world view at 60 m; source records and datasets remain available below.';
    document.body.dataset.waterReady='fallback';
  }
}
function save(replace=false) {
  const params=new URLSearchParams({region:state.region,depth:String(state.depth)});
  history[replace?'replaceState':'pushState']({},'',`${location.pathname}?${params}${location.hash}`);
}
async function start() {
  try {
    const r=await fetch('data/submerged/grids.json');if(!r.ok)throw Error();manifest=await r.json();
    state=readWaterState(location.search,manifest.grids);$('#water-tools').hidden=false;
    await loadRegion();
  } catch {$('#water-status').textContent='Interactive data could not load. The static world maps above show the baseline and a 60 m depth screen; all records remain readable.';return;}
  $('#water-depth').addEventListener('input',e=>{state.depth=+e.target.value;save(true);draw();});
  document.addEventListener('click',e=>{
    const b=e.target.closest('[data-water-region]'),link=e.target.closest('[data-region-link]');
    if(link&&(e.metaKey||e.ctrlKey||e.shiftKey||e.altKey))return;
    const region=b?.dataset.waterRegion||link?.dataset.regionLink;if(!region)return;
    if(link)e.preventDefault();state.region=region;save();loadRegion();
    if(link)$('#landscape').scrollIntoView();
  });
  window.addEventListener('popstate',()=>{state=readWaterState(location.search,manifest.grids);loadRegion();});
}
start();
